### ISSUE TITLE
Feature Request: Differential Evolution MCMC

### ISSUE THREAD

**Issue by itsdfish** [TuringLang/AdvancedPS.jl#9]
Differential evolution MCMC is a particle sampler that works well for ABC and other situations in which AD is not feasible.  Although it is less efficient than NUTS, it performs better with correlated/irregular posterior distributions compared to many other alternatives. I think it would be a useful addition to Turing. I have a naive implementation in [DifferentialEvolutionMCMC](https://github.com/itsdfish/DifferentialEvolutionMCMC.jl) that partially follows the AbstractMCMC template. Nonetheless, I think it would benefit from the expertise of Turing developers, who could integrate it better with the Turing ecosystem, make it composable,  and optimize performance. If anyone is interested in this algorithm, feel free to reuse any code in DifferentialEvolutionMCMC. 

References:
[Approximate Bayesian computation with differential evolution
](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.721.9736&rep=rep1&type=pdf)

**Comment by cpfiffer** [comment]
I'd be interested in seeing this added to Turing's sampler base, but I think a lot of the Turing people are quite busy for the moment and are probably unable to integrate this. If you want to take a crack at it we're happy to give some pointers.

**Comment by itsdfish** [comment]
Understandable. 

Currently, I pass a sampler object to a `sample` method and it returns a chain, just like other algorithms that use Turing. The main problem I am having is integrating it with the `@model ` macro. Is there a relatively simple example of this that you could direct me to?

**Comment by cpfiffer** [comment]
I think the best so far is the Turing/EllipticalSliceSampling integration [here](https://github.com/TuringLang/Turing.jl/blob/master/src/inference/ess.jl).

**Comment by itsdfish** [comment]
Thank you. I'll try to make sense out of it. Is the basic idea that each sampler/sampler family has a file in inference with various methods (such as those defined in ESS.jl)?

**Comment by cpfiffer** [comment]
Yeah, that's more or less the idea, though I think we want general plug-ins to go into `src/contrib/inference`.

**Comment by torfjelde** [comment]
First off, cool stuff @itsdfish !
Second, probably better to make a PR to Turing directly rather than AdvancedPS, given the current state of AdvancedPS?

**Comment by itsdfish** [comment]
Thanks, @torfjelde. I should have sometime during the weekend to start looking into Turing's interface. I'll can submit a PR to Turing when it is ready or I reach an impasse.

**Comment by torfjelde** [comment]
Awesome:) Feel free to hit us up if there's any questions!

**Comment by devmotion** [comment]
Just came across the issue and the discussion here :smiley: 

We're preparing some drastic changes of the AbstractMCMC internals (see https://github.com/TuringLang/AbstractMCMC.jl/pull/42) (the user-facing API is unchanged), so IMO it would make sense if you could already target this upcoming release (I'll check your code, maybe I can make a simple PR similar to https://github.com/TuringLang/EllipticalSliceSampling.jl/blob/abstractmcmc2/src/abstractmcmc.jl). BTW an implementation of the AbstractMCMC interface is not a requirement for providing Turing support for your sampler - e.g., IIRC initially the elliptical slice sampler in EllipticalSliceSampling did not implement the interface and we just called into whatever API was used back then.

In general, I'm not sure if it's actually useful to gather all the files in the Turing repo for samplers in other repos. IMO it's quite likely that these integrations become outdated at some point, and also it leads to all kind of redundancies. I think in the long term a nice approach would be if these external packages (IMO it's fine to keep stuff in clearly separated, (hopefully) well-maintained packages) could implement the AbstractMCMC interface for their sampler with models of type `DynamicPPL.Model` (they could either depend on DynamicPPL explicitly - the dependencies don't seem to be too bad in https://github.com/TuringLang/DynamicPPL.jl/blob/275ccc8791be7d5dc4eb74b6e0c19de247f56d74/Project.toml#L7-L12, I guess Bijectors is pulling in quite some implicit dependencies though - or put it in a Requires block - which increases loading times as well though...). A natural approach would to just forward the required implementations of the interface to calls of the same methods of the AbstractMCMC with their own custom model types - basically to methods such as the ones defined in https://github.com/TuringLang/EllipticalSliceSampling.jl/blob/abstractmcmc2/src/abstractmcmc.jl.

**Comment by itsdfish** [comment]
Thanks @devmotion. I am currently leaning towards keeping the overloaded functions because given the structure of Differential Evolution, the current code is easier for me to understand and maintain. Nonetheless, it is still useful because the user experience is much the same. Having this consistency across packages makes it easier for users to reuse knowledge and design patterns. 

I do like your idea to keep samplers in packages that are separate from Turing.  It makes it cleaner and easier to maintain. Plus, if a user only has to call using Turing  and using PreferredSampler, perhaps it could make load times faster, or at least make Turing less bulky. Is there currently a way to do this or is this a feature that you intend to add in the future?

**Comment by devmotion** [comment]
>  It makes it cleaner and easier to maintain. Plus, if a user only has to call using Turing and using PreferredSampler, perhaps it could make load times faster, or at least make Turing less bulky. 

I agree.

> Is there currently a way to do this or is this a feature that you intend to add in the future?

Not a "nice" way. You could just add an optional dependency on Turing, hide everything with Requires, and then include basically the same file that you would submit to Turing.jl in your own repo. But I guess it would be better to do this in a more standardized way. I have some ideas for how to achieve this (e.g., by changing the implementation of `Sampler` in DynamicPPL such that it just wraps the sampling algorithm (which could be defined in an external package) and the sample space (i.e., the variables of the Turing model for which the algorithm should be used) since we also need the sample space in Turing), but I guess it's best to first update everything to AbstractMCMC 2 before making any additional changes.

**Comment by itsdfish** [comment]
That sounds great. Would you recommend waiting for these changes to be implemented before I try to interface with Turing?

**Comment by devmotion** [comment]
Depends on how urgent you need the Turing support, I guess. It might take a while before all changes are in place and propagated through the whole ecosystem. I'll also be on vacation for some weeks, so I can't provide a good time estimate right now. In any case, it shouldn't be difficult to update it later on - the advantage of waiting for the changes would just be that probably the integration would be much easier or even (almost) for free.

**Comment by itsdfish** [comment]
Thanks for the info. I'll wait for the changes before I provide Turing support.

**Comment by pcjentsch** [comment]
worth noting that the package KissABC contains an implementation of this method
